scripts/scmigrate/migrate_test.go

```diff
@@ -0,0 +1,172 @@
+package scmigrate
+
+import (
+	"context"
+	"math/rand"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/tendermint/tendermint/types"
+)
+
+func appendRandomMigrations(in []toMigrate, num int) []toMigrate {
+	if in == nil {
+		in = []toMigrate{}
+	}
+
+	for i := 0; i < num; i++ {
+		height := rand.Int63()
+		if height <= 0 {
+			continue
+		}
+		in = append(in, toMigrate{commit: &types.Commit{Height: height}})
+	}
+	return in
+}
+
+func assertWellOrderedMigrations(t *testing.T, testData []toMigrate) {
+	t.Run("ValuesDescend", func(t *testing.T) {
+		for idx := range testData {
+			height := testData[idx].commit.Height
+			if idx == 0 {
+				continue
+			}
+			prev := testData[idx-1].commit.Height
+			if prev < height {
+				t.Fatal("height decreased in sort order")
+			}
+		}
+	})
+	t.Run("EarliestIsZero", func(t *testing.T) {
+		earliestHeight := testData[len(testData)-1].commit.Height
+		if earliestHeight != 0 {
+			t.Fatalf("the earliest height is not 0: %d", earliestHeight)
+		}
+	})
+}
+
+func insertTestData(t *testing.T, db dbm.DB, data []toMigrate) {
+	t.Helper()
+
+	batch := db.NewBatch()
+
+	for idx, val := range data {
+		payload, err := proto.Marshal(val.commit.ToProto())
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if err := batch.Set(makeKeyFromPrefix(prefixSeenCommit, int64(idx)), payload); err != nil {
+			t.Fatal(err)
+		}
+	}
+	if err := batch.WriteSync(); err != nil {
+		t.Fatal(err)
+	}
+	if err := batch.Close(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestMigrations(t *testing.T) {
+	t.Run("Sort", func(t *testing.T) {
+		t.Run("HandCraftedData", func(t *testing.T) {
+			testData := []toMigrate{
+				{commit: &types.Commit{Height: 100}},
+				{commit: &types.Commit{Height: 0}},
+				{commit: &types.Commit{Height: 8}},
+				{commit: &types.Commit{Height: 1}},
+			}
+
+			sortMigrations(testData)
+			assertWellOrderedMigrations(t, testData)
+		})
+		t.Run("RandomGeneratedData", func(t *testing.T) {
+			testData := []toMigrate{{commit: &types.Commit{Height: 0}}}
+
+			testData = appendRandomMigrations(testData, 10000)
+
+			sortMigrations(testData)
+			assertWellOrderedMigrations(t, testData)
+		})
+	})
+	t.Run("KeyProducer", func(t *testing.T) {
+		defer func() {
+			if r := recover(); r == nil {
+				t.Fatal("operation should panic")
+			}
+
+		}()
+
+		// should panic if ints rather than int64s
+		makeKeyFromPrefix(int64(3), int(3))
```

Comment:
If we update the signature to require `int64`, this will be statically impossible.